animation_nodes/algorithms/mesh_generation/find_shortest_path.pyx

```diff
@@ -0,0 +1,112 @@
+from ... data_structures import GPStroke
+from ... data_structures cimport (
+    Mesh,
+    LongList,
+    ColorList,
+    FloatList,
+    PolySpline,
+    DoubleList,
+    BooleanList,
+    Vector3DList,
+)
+
+from libc.math cimport INFINITY
+from . line import getLinesMesh
+from ... math cimport distanceVec3
+
+# Dijkstra's algorithm (https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm) but is implemented
+# such a way to handle multi-sources and mesh with multiple islands.
+def getShortestPath(Mesh mesh, LongList sources, LongList targets, str pathType, str mode):
+    cdef Vector3DList vertices = mesh.vertices
+    cdef long vertexCount = vertices.length
+    cdef float maxDistance = INFINITY
+    cdef DoubleList distances = DoubleList.fromValue(maxDistance, length = vertexCount)
+    cdef LongList previousVertices = LongList.fromValue(-1, length = vertexCount)
+    cdef BooleanList visitedVertices = BooleanList.fromValue(False, length = vertexCount)
+    cdef Py_ssize_t i, targetIndex
+
+    for i in range (sources.length):
+        distances.data[sources.data[i]] = 0.0
+
+    if mode == "PATH":
+        targetIndex = targets.data[0]
+    else:
+        targetIndex = <int>INFINITY
+
+    cdef float distance, minDistance
+    cdef Py_ssize_t k, l, currentIndex, neighboursStart, neighbourIndex
+    cdef LongList neighboursAmounts, neighboursStarts, neighbours
+    neighboursAmounts, neighboursStarts, neighbours = mesh.getLinkedVertices()[:3]
+
+    for i in range(vertexCount):
+
+        minDistance = maxDistance
+        for k in range(vertexCount):
+            distance = distances.data[k]
+            if distance < minDistance and not visitedVertices.data[k]:
+                minDistance = distance
+
+        if minDistance == maxDistance:
+            break
+
+        for currentIndex in range(vertexCount):
+            if abs(minDistance - distances.data[currentIndex]) < 1.0e-6:
+                visitedVertices.data[currentIndex] = True
```

Comment:
If you want the index of the vertex with the smallest distance, you compute it directly, don't do two nested loops. Every time you assign `minDistance = distance` you also assign `k` to an index variable.

Reply:
Omar, we need to compute *index of the vertex with the smallest distance* which is not visited yet.

Reply:
Isn't this why we have `not visitedVertices.data[k]:`?

Reply:
Yes.

Reply:
So what is the difficulty?

Reply:
I do not how to compute it directly.

Reply:
```cython
        minDistance = maxDistance
        cdef Py_ssize_t minDistanceIndex 
        for k in range(vertexCount):
            distance = distances.data[k]
            if not visitedVertices.data[k] and distance < minDistance:
                minDistanceIndex = k
                minDistance = distance

```

Reply:
Okay. Omar this is possible when there is the only one source. If we have multiple sources then we can have then multiple vertices that can have the same `minDistance` value. They generate different results:
**With Loop**
![With_Loop](https://user-images.githubusercontent.com/30294746/93346322-7a881700-f851-11ea-9dbd-904687ddb658.png)

**With minDistanceIndex**
![With_Index](https://user-images.githubusercontent.com/30294746/93346367-870c6f80-f851-11ea-806b-5e8056e06fcf.png)

Reply:
Omar, I have tested both methods for different cases, path topology is different but path length is the same.

Reply:
So it is different but correct? Are you saying we shouldn't use the index method?

Reply:
**With Loop**: In this method, we are updating the distance (and previous index) of all `min distance` vertices which lies in region/reach of each source. It is like parallel update. 

**With Index**: In this method, we are updating the distance (and previous index) of the first `min distance` vertex. It is like series update.

The difference in topology is mainly pronouced when base mesh has edges with same length e.g., Grid Mesh.

Reply:
Notice that the loop is slower. So if the difference in topology is not important, we should use the index method.

Reply:
Yes. I agree with you.